x/evm/types/statedb.go

```diff
@@ -702,38 +702,39 @@ func (csdb *CommitStateDB) Copy() *CommitStateDB {
 		storeKey:             csdb.storeKey,
 		paramSpace:           csdb.paramSpace,
 		accountKeeper:        csdb.accountKeeper,
-		stateObjects:         make([]stateEntry, len(csdb.journal.dirties)),
-		addressToObjectIndex: make(map[ethcmn.Address]int, len(csdb.journal.dirties)),
-		stateObjectsDirty:    make(map[ethcmn.Address]struct{}, len(csdb.journal.dirties)),
+		stateObjects:         []stateEntry{},
+		addressToObjectIndex: make(map[ethcmn.Address]int),
+		stateObjectsDirty:    make(map[ethcmn.Address]struct{}),
 		refund:               csdb.refund,
 		logSize:              csdb.logSize,
 		preimages:            make(map[ethcmn.Hash][]byte),
 		journal:              newJournal(),
 	}
 
 	// copy the dirty states, logs, and preimages
-	for i, dirty := range csdb.journal.dirties {
+	for _, dirty := range csdb.journal.dirties {
 		// There is a case where an object is in the journal but not in the
 		// stateObjects: OOG after touch on ripeMD prior to Byzantium. Thus, we
 		// need to check for nil.
 		//
 		// Ref: https://github.com/ethereum/go-ethereum/pull/16485#issuecomment-380438527
 		if idx, exist := csdb.addressToObjectIndex[dirty.address]; exist {
-			state.stateObjects[i] = stateEntry{
+			state.stateObjects = append(state.stateObjects, stateEntry{
 				address:     dirty.address,
 				stateObject: csdb.stateObjects[idx].stateObject.deepCopy(state),
-			}
-			state.stateObjectsDirty[dirty.address] = struct{}{}
+			})
+			delete(state.stateObjectsDirty, dirty.address)
+			state.addressToObjectIndex[dirty.address] = len(state.stateObjects) - 1
```

Comment:
this is the fix, right?

Reply:
unfortunately not, the fix seemed to be `idx < len(csdb.stateObjects)` inside `getStateObject` and `setStateObject` which I don't really like...